Dockerfile

```diff
@@ -1,20 +1,64 @@
-# Build Chainlink
-FROM golang:1.10-alpine as builder
+# Build Chainlink with SGX
+FROM smartcontract/builder as builder
 
-RUN apk add --no-cache make curl git gcc g++ musl-dev linux-headers yarn python
-RUN curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
+# Have to reintroduce ENV vars from Baidu's SGX SDK image
+ENV PATH /root/.cargo/bin:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/sgxsdk/bin:/opt/sgxsdk/bin/x64
+ENV LD_LIBRARY_PATH /opt/sgxsdk/sdk_libs
+ENV SGX_SDK /opt/sgxsdk
+
+ARG COMMIT_SHA
+ARG ENVIRONMENT
+ARG SGX_ENABLED
+ARG SGX_SIMULATION
 
 ADD . /go/src/github.com/smartcontractkit/chainlink
-RUN cd /go/src/github.com/smartcontractkit/chainlink && make build
+WORKDIR /go/src/github.com/smartcontractkit/chainlink
+RUN make build
+
+# Final layer: ubuntu with aesm and chainlink binaries (executable + enclave)
+FROM ubuntu:16.04
+
+# Install AESM
+RUN apt-get update && \
+  apt-get install -y \
+    curl \
+    libssl-dev \
+    libcurl4-openssl-dev \
+    libprotobuf-dev \
+    kmod \
+    libprotobuf-c0-dev \
+    libxml2-dev \
+    libssl1.0.0
+
+RUN /usr/sbin/useradd aesmd 2>/dev/null
+
+RUN mkdir -p /var/opt/aesmd && chown aesmd.aesmd /var/opt/aesmd
+RUN mkdir -p /var/run/aesmd && chown aesmd.aesmd /var/run/aesmd
 
-# Copy Chainlink into a second stage deploy alpine container
-FROM alpine:latest
+COPY --from=builder /opt/sgxsdk/lib64/libsgx*.so /usr/lib/
+COPY --from=builder /opt/intel/ /opt/intel/
 
-RUN apk add --no-cache ca-certificates
-COPY --from=builder /go/src/github.com/smartcontractkit/chainlink/chainlink /usr/local/bin/
-COPY --from=builder /go/src/github.com/smartcontractkit/chainlink/gui/dist /go/src/github.com/smartcontractkit/chainlink/gui/dist
+# Copy chainlink enclave+stub from build image
 
+# XXX: This is structured like so because wheb SGX_ENABLED is 'no', no sgx
+# targets get produced, so here we specify the SGX targets as a wildcard with
+# chainlink so COPY doesn't fail when they're missing. Then move them all into
+# place.
+ARG ENVIRONMENT
+COPY --from=builder \
+  /go/src/github.com/smartcontractkit/chainlink/chainlink \
+  /go/src/github.com/smartcontractkit/chainlink/sgx/target/$ENVIRONMENT/*.so \
+  /tmp
+RUN mv /tmp/libadapters.so /usr/lib 2>/dev/null || \
+  mv /tmp/enclave.signed.so /root 2>/dev/null || \
+  rm /tmp/*.so 2>/dev/null || \
+  rm /tmp/chainlink
 
+# Launch chainlink via a small script that watches AESM + Chainlink
+ARG SGX_SIMULATION
+ENV SGX_SIMULATION $SGX_SIMULATION
+WORKDIR /root
+COPY ./chainlink-launcher.sh /root
+RUN chmod +x ./chainlink-launcher.sh
 EXPOSE 6688
-EXPOSE 6689
-ENTRYPOINT ["chainlink"]
+ENTRYPOINT ["./chainlink-launcher.sh"]
```

Comment:
Do we always want the chainlink-launcher to be the entry point?

Reply:
The chainlink launcher should be functionally equivalent whether it's running in SGX mode or not. If we don't like that, what we could do is build a different set of scripts in the build stage and move the one we'll use into place. We'd likely only use the chainlink-launcher script for actual SGX hardware though, since AESM doesn't run in emulation mode.